tests/frame.cpp

```diff
@@ -0,0 +1,333 @@
+  for (int i = 0; i < nThreads; ++i) {
+    threads.emplace_back([&frame, i]() {
+      auto clusters = ExampleClusterCollection();
+      clusters.create(i * 3.14);
+      clusters.create(i * 3.14);
+      frame.put(std::move(clusters), "clusters_" + std::to_string(i));
+
+      auto hits = ExampleHitCollection();
+      hits.create(i * 100ULL);
+      hits.create(i * 100ULL);
+      hits.create(i * 100ULL);
+      frame.put(std::move(hits), "hits_" + std::to_string(i));
+    });
+  }
+
+  for (auto& t : threads) {
+    t.join();
+  }
+
+  // Check the frame contents after all threads have finished
+  for (int i = 0; i < nThreads; ++i) {
+    auto& hits = frame.get<ExampleHitCollection>("hits_" + std::to_string(i));
+    REQUIRE(hits.size() == 3);
+    for (const auto h : hits) {
+      REQUIRE(h.cellID() == i * 100ULL);
+    }
+
+    auto& clusters = frame.get<ExampleClusterCollection>("clusters_" + std::to_string(i));
+    REQUIRE(clusters.size() == 2);
+    for (const auto c : clusters) {
+      REQUIRE(c.energy() == i * 3.14);
+    }
+  }
+}
+
+// Helper function to create a frame in the tests below
+auto createFrame() {
+  auto frame = podio::Frame();
+
+  frame.put(ExampleClusterCollection(), "emptyClusters");
+
+  // Create a few hits inline (to avoid having to have two identifiers)
+  auto& hits = frame.put(
+      []() {
+        auto coll = ExampleHitCollection();
+        auto hit = coll.create(0x42ULL, 0., 0., 0., 0.);
+        auto hit2 = coll.create(0x123ULL, 1., 1., 1., 1.);
+        return coll;
+      }(),
+      "hits");
+
+  auto clusters = ExampleClusterCollection();
+  auto cluster = clusters.create(3.14f);
+  cluster.addHits(hits[0]);
+  auto cluster2 = clusters.create(42.0f);
+  cluster2.addHits(hits[1]);
+  cluster2.addClusters(cluster);
+
+  // Create a few clustes inline and relate them to the hits from above
+  frame.put(std::move(clusters), "clusters");
+
+  frame.putParameter("anInt", 42);
+  frame.putParameter("someFloats", {1.23f, 2.34f, 3.45f});
+
+  return frame;
+}
+
+// Helper function to get names easily below
+std::string makeName(const std::string& prefix, int index) {
+  return prefix + "_" + std::to_string(index);
+}
+
+TEST_CASE("Frame collections multithreaded insert and read", "[frame][basics][multithread]") {
+  constexpr int nThreads = 10;
+  std::vector<std::thread> threads;
+  threads.reserve(10);
+
+  // create a pre-populated frame
+  auto frame = createFrame();
+
+  // Fill collections from different threads
+  for (int i = 0; i < nThreads; ++i) {
+    threads.emplace_back([&frame, i]() {
+      auto clusters = ExampleClusterCollection();
+      clusters.create(i * 3.14);
+      clusters.create(i * 3.14);
+      frame.put(std::move(clusters), makeName("clusters", i));
+
+      // Retrieve a few collections in between and do iust a very basic testing
+      auto& existingClu = frame.get<ExampleClusterCollection>("clusters");
+      REQUIRE(existingClu.size() == 2);
```

Comment:
catch2 macros are not yet threadafe:
https://github.com/catchorg/Catch2/blob/devel/docs/limitations.md#thread-safe-assertions